ehrql/serializer.py

```diff
@@ -0,0 +1,265 @@
+import datetime
+import functools
+import json
+import pathlib
+
+from ehrql.codes import BaseCode
+from ehrql.file_formats.base import BaseDatasetReader
+from ehrql.measures.measures import Measure
+from ehrql.query_language import DummyDataConfig
+from ehrql.query_model.column_specs import ColumnSpec
+from ehrql.query_model.nodes import Node, Position, Value
+from ehrql.query_model.table_schema import BaseConstraint, Column, TableSchema
+from ehrql.utils.module_utils import get_all_subclasses
+
+
+# Mapping of names to types for all types that we support serializing/deserializing
+TYPE_REGISTRY = {
+    type_.__qualname__: type_
+    for type_ in [
+        bool,
+        int,
+        float,
+        str,
+        datetime.date,
+        tuple,
+        dict,
+        frozenset,
+        Position,
+        TableSchema,
+        Column,
+        pathlib.Path,
+        ColumnSpec,
+        DummyDataConfig,
+        Measure,
+        *get_all_subclasses(Node),
+        *get_all_subclasses(BaseCode),
+        *get_all_subclasses(BaseConstraint),
+        *get_all_subclasses(BaseDatasetReader),
+    ]
+}
+
+TYPE_REGISTRY_INVERSE = {type_: name for name, type_ in TYPE_REGISTRY.items()}
+
+# Ensure no name clashes
+assert len(TYPE_REGISTRY) == len(TYPE_REGISTRY_INVERSE)
+
+
+def type_name(obj):
+    return TYPE_REGISTRY_INVERSE[type(obj)]
+
+
+def serialize(value):
+    return json.dumps(Marshaller.to_dict(value), indent=2)
+
+
+def deserialize(data):
+    return Unmarshaller.from_dict(json.loads(data))
+
+
+class Marshaller:
+    """
+    Convert arbitrary values to nested structures of JSON-supported values ready to be
+    serialized
+    """
+
+    @classmethod
+    def to_dict(cls, value):
+        marshaller = cls()
+        return {
+            "value": marshaller.marshal(value),
+            **marshaller.get_references(),
+        }
+
+    def __init__(self):
+        self.references = {}
+        # Sequence of unique IDs (strings not ints so they can be used as JSON keys)
+        id_sequence = (str(i) for i in range(1, 2**32))  # pragma: no branch
+        self.get_next_reference_id = id_sequence.__next__
+
+    def get_references(self):
+        return dict(self.references.values())
+
+    @functools.singledispatchmethod
+    def marshal(self, obj):
+        assert False, f"Unsupported type '{type(obj)}': {obj!r}"
+
+    @marshal.register(type(None))
+    @marshal.register(bool)
+    @marshal.register(str)
+    @marshal.register(int)
+    @marshal.register(float)
+    def marshal_primitive(self, obj):
+        return obj
```

Comment:
I think it is desirable from the point of view of producing sensible looking output. It's the same consideration that motivate treating `Value` differently. It means we can serialize `Value(10)` as:
```
{"Value": 10}
```
rather than as
```
{"Value": {"value": {"int": 10}}}
```

Reply:
On this, and the other stylistic (though not _merely_ stylistic) considerations, I'm not dismissing your viewpoint at all but I'd suggest deferring the discussion as long what I've got currently isn't absolutely terrible. It's all very easy to change at this stage given it's effectively a private IPC mechanism.

At some point we'll have a need to make this part of the public API surface and at _that_ point we'll need to do some serious thinking which should involve more than just the two of us. Does that sound reasonable?

Reply:
Yes, totally.